source/agora/consensus/EnrollmentPool.d

```diff
@@ -0,0 +1,404 @@
+
+        // create the table for enrollment pool if it doesn't exist yet
+        this.db.execute("CREATE TABLE IF NOT EXISTS enrollment_pool " ~
+            "(key TEXT PRIMARY KEY, val BLOB NOT NULL)");
+    }
+
+    /***************************************************************************
+
+        Shut down the database
+
+        Note: this method must be called explicitly, and not inside of
+        a destructor.
+
+    ***************************************************************************/
+
+    public void shutdown ()
+    {
+        this.db.close();
+    }
+
+    /***************************************************************************
+
+        Add a enrollment data to the enrollment pool
+
+        Params:
+            enroll = the enrollment data to add
+            finder = the delegate to find UTXOs with
+
+        Returns:
+            true if the enrollment data has been added to the enrollment pool
+
+    ***************************************************************************/
+
+    public bool add (const ref Enrollment enroll, scope UTXOFinder finder)
+        @safe nothrow
+    {
+        // check validity of the enrollment data
+        if (auto reason = isInvalidEnrollmentReason(enroll, finder))
+        {
+            log.info("Invalid enrollment data: {}, Data was: {}", reason, enroll);
+            return false;
+        }
+
+        // check if already exists
+        if (this.hasEnrollment(enroll.utxo_key))
+        {
+            log.info("Rejected already existing enrollment, Data was: {}",
+                enroll);
+            return false;
+        }
+
+        static ubyte[] buffer;
+        try
+        {
+            serializeToBuffer(enroll, buffer);
+        }
+        catch (Exception ex)
+        {
+            log.error("Serialization error: {}, Data was: {}", ex.msg, enroll);
+            return false;
+        }
+
+        try
+        {
+            () @trusted {
+                this.db.execute("INSERT INTO enrollment_pool (key, val) VALUES (?, ?)",
+                    enroll.utxo_key.toString(), buffer);
+            }();
+        }
+        catch (Exception ex)
+        {
+            log.error("Database operation error: {}, Data was: {}", ex.msg, enroll);
+            return false;
+        }
+
+        return true;
+    }
+
+    /***************************************************************************
+
+        Returns:
+            the number of enrollments being managed by the enrollment pool,
+            which are un-registered enrollments.
+
+    ***************************************************************************/
+
+    public size_t count () @trusted nothrow
+    {
+        size_t cnt;
+        try
+        {
+            cnt = this.db.execute("SELECT count(*) FROM enrollment_pool").
+                oneValue!size_t;
+        }
+        catch (Exception ex)
+        {
+            log.error("Database operation error on count");
+        }
```

Comment:
```suggestion
        try
        {
            return this.db.execute("SELECT count(*) FROM enrollment_pool").
                oneValue!size_t;
        }
        catch (Exception ex)
        {
            log.error("Database operation error on count");
            return 0;
        }
```

And the documentation needs to be updated to say 0 is returned in case of errors.